### PR TITLE
Add object locking around updates + deletes

### PR DESCRIPTION
**Pull request by cbeer**
Or maybe there's something else in the architecture that's stopping simultaneous updates?

### CODE REVIEW

**Comment by justinlittman** [comment]
Will second attempt to lock block or raise? (I don't love blocking, as it might lock up http requests. I'd prefer to see it raise so that the caller can retry.)

**Comment by cbeer** [comment]
We can make it do whatever. It's not yet clear to me whether this is needed (but also not clear that it isn't.). What do you think?

**Comment by justinlittman** [comment]
In theory it shouldn't happen. However, since on the DSA side publishing is performed by a job, anything is possible given retries. This seems lightweight enough that it is worth it.

File: app/services/versioned_files_service/migrate_action.rb [hunk 1]
```diff
@@ -9,26 +9,34 @@ def initialize(object:, version_metadata:)
...
-      # For each shelved file, create a hardlink named by md5
-      link_content_files
-      # Write the cocina to cocina path for the version and create a new head cocina symlink.
-      write_cocina(version: 1, cocina: cocina_hash, head_version: true)
-      # Write the public xml to public xml path for the version and create a new head public xml symlink .
-      write_public_xml(version: 1, public_xml:, head_version: true)
-      copy_meta_json
-      # Update the version manifest.
-      version_manifest.update_version(version: 1, version_metadata:, head_version: true)
+      create_directory_structure
```
Comment (justinlittman) [inline app/services/versioned_files_service/migrate_action.rb:12]: I don't think we should create the versions directory before checking the content files. Once the versions directory is created, applications will think that the object is using the versioned layout, even though the directory is empty.

Comment (justinlittman) [reply app/services/versioned_files_service/migrate_action.rb:12]: If you just switch `check_content_files!` and `create_directory_structure` I think the logic should be fine.

Comment (cbeer) [reply app/services/versioned_files_service/migrate_action.rb:12]: Except the lock file is created within the AWFL directory structure?

**Review by justinlittman — APPROVED** [review]

**Pull request merged**